DeepCopy should generate types within k8s.io/kubernetes

smarterclayton opened the pull request:
Take an incremental step to control which types get DeepCopy generated.

Fixes #26036, #24667

@thockin @wojtek-t @lavalamp

smarterclayton committed:
DeepCopy should only generate types in k8s.io/kubernetes

Also make generation more selective (require opt in) to prevent lots of
unnecessary generated deep copies.

smarterclayton commented:
```
Kubernetes master is running at https://104.155.131.166
GLBCDefaultBackend is running at https://104.155.131.166/api/v1/proxy/namespaces/kube-system/services/default-http-backend
Elasticsearch is running at https://104.155.131.166/api/v1/proxy/namespaces/kube-system/services/elasticsearch-logging
Heapster is running at https://104.155.131.166/api/v1/proxy/namespaces/kube-system/services/heapster
Kibana is running at https://104.155.131.166/api/v1/proxy/namespaces/kube-system/services/kibana-logging
KubeDNS is running at https://104.155.131.166/api/v1/proxy/namespaces/kube-system/services/kube-dns
kubernetes-dashboard is running at https://104.155.131.166/api/v1/proxy/namespaces/kube-system/services/kubernetes-dashboard
Grafana is running at https://104.155.131.166/api/v1/proxy/namespaces/kube-system/services/monitoring-grafana
InfluxDB is running at https://104.155.131.166/api/v1/proxy/namespaces/kube-system/services/monitoring-influxdb

Failed to create firewall e2e-gce-jenkins-pr-agent-20160523-2-2-minion-e2e-gce-jenkins-pr-agent-20160523-2-2-http-alt in kubernetes-jenkins-pull
2016/05/24 10:28:28 e2e.go:206: Error running up: exit status 1
2016/05/24 10:28:28 e2e.go:202: Step 'up' finished in 24m18.361443677s
2016/05/24 10:28:28 e2e.go:114: Error starting e2e cluster. Aborting.
exit status 1
+ dump_cluster_logs_and_exit
+ local -r exit_status=1
+ [[ -x cluster/log-dump.sh ]]
+ ./cluster/log-dump.sh /var/lib/jenkins/workspace/kubernetes-pull-build-test-e2e-gce/_artifacts
Dumping master and node logs to /var/lib/jenkins/workspace/kubernetes-pull-build-test-e2e-gce/_artifacts
Warning: Permanently added '104.155.131.166' (ECDSA) to the list of known hosts.
Warning: Permanently added '104.154.46.254' (ECDSA) to the list of known hosts.
Warning: Permanently added '130.211.204.24' (ECDSA) to the list of known hosts.
Warning: Permanently added '104.197.67.30' (ECDSA) to the list of known hosts.
Warning: Permanently added '23.251.146.166' (ECDSA) to the list of known hosts.
Warning: Permanently added '104.155.135.177' (ECDSA) to the list of known hosts.
Warning: Permanently added '104.154.238.203' (ECDSA) to the list of known hosts.
+ [[ true == \t\r\u\e ]]
+ go run ./hack/e2e.go -v --down
2016/05/24 10:28:34 e2e.go:200: Running: teardown
Project: kubernetes-jenkins-pull
Zone: us-central1-f
Shutting down test cluster in background.
ERROR: (gcloud.compute.firewall-rules.delete) Some requests did not succeed:
 - Invalid value 'e2e-gce-jenkins-pr-agent-20160523-2-2-minion-e2e-gce-jenkins-pr-agent-20160523-2-2-http-alt'. Values must match the following regular expression: '[a-z](?:[-a-z0-9]{0,61}[a-z0-9])?'

ERROR: (gcloud.compute.firewall-rules.delete) Some requests did not succeed:
 - Invalid value 'e2e-gce-jenkins-pr-agent-20160523-2-2-minion-e2e-gce-jenkins-pr-agent-20160523-2-2-nodeports'. Values must match the following regular expression: '[a-z](?:[-a-z0-9]{0,61}[a-z0-9])?'

Bringing down cluster using provider: gcc
```

smarterclayton commented:
@k8s-bot test this issue #26187

thockin commented:
I will take a look later today.

On Tue, May 24, 2016 at 11:35 AM, Clayton Coleman <notifications@github.com>
wrote:

> @k8s-bot <https://github.com/k8s-bot> test this issue #26187
> <https://github.com/kubernetes/kubernetes/issues/26187>
>
> —
> You are receiving this because you were mentioned.
> Reply to this email directly or view it on GitHub
> <https://github.com/kubernetes/kubernetes/pull/26179#issuecomment-221356242>
>

smarterclayton commented:
Still needs review.

smarterclayton commented:
@wojtek-t can you review

wojtek-t commented:
@smarterclayton - sorry, yesterday was holiday in Poland and I was super busy this week.

Looking now.

wojtek-t commented on pkg/api/meta/metatypes/deep_copy_generated.go:
```diff
@@ -1,33 +0,0 @@
-// +build !ignore_autogenerated
```
Don't we want to generate this one?

smarterclayton replied:
It does not appear to be used in any path that does deep copies
(instead, it's copied by the garbage collector when it calls
meta.Accessor).  I think it's safe to remove.

I'm going to in a follow on print debug info if we copy something
without generated structs.  However, the reflect copy path should be
fast for this type.

wojtek-t commented:
Single very minor nit.

LGTM - feel free to self-apply lgtm label after answering comment.

thockin commented:
I'm removing LGTM because I need to process this.  I don't like adding syntax for this, but I need to understand what you're doing.  My #25978 (don't check in generated code) solves this very differently.

I added a new "-x" flag to the generators which is "extra dirs".  Extra dirs are parsed and loaded but not considered as inputs.  I also added a --self-register flag to deep-copy.  Then I used per-package tag comments to drive which packages get generated.

Before this (your) PR goes in I want to really understand it because it could toally hork my PR, and I have put a lot of work into it (and it mostly works!!).

smarterclayton commented:
-packages are -x (more or less), +packages are not register.  I don't
particularly care about the syntax (I'm fine with your approach), although
gencopy=true on a -package is good for limiting copies on meaningless types
(all the previous gorp in pkg/runtime that is never deep copied).  I don't
think we want to generate copies for all structs in all packages (at least
for now, until we split the serializable types out of pkg/runtime).

The package constraints is probably still important - it controls the
package set that can be considered input, but if you have that equivalent
due to per packages and controlling which packages you scan (anything
outside the k8s space, or in Openshift or contrib anything inside that
space).

I'm fine with the proposed approach, as long as we can make generation opt
in as necessary.

One change I realized we need to make is to skip generation of methods
which already exist (there is a more efficient copy in
pkg/api/resource/deep_copy.go that should override the generated one).
I'll do that once we sort out the above discussion.

On May 28, 2016, at 5:07 PM, Tim Hockin <notifications@github.com> wrote

I'm removing LGTM because I need to process this. I don't like adding
syntax for this, but I need to understand what you're doing. My #25978
<https://github.com/kubernetes/kubernetes/pull/25978> (don't check in
generated code) solves this very differently.

I added a new "-x" flag to the generators which is "extra dirs". Extra dirs
are parsed and loaded but not considered as inputs. I also added a
--self-register flag to deep-copy. Then I used per-package tag comments to
drive which packages get generated.

Before this (your) PR goes in I want to really understand it because it
could toally hork my PR, and I have put a lot of work into it (and it
mostly works!!).

—
You are receiving this because you were mentioned.
Reply to this email directly, view it on GitHub
<https://github.com/kubernetes/kubernetes/pull/26179#issuecomment-222329746>,
or mute the thread
<https://github.com/notifications/unsubscribe/ABG_pzP1iOM2MbIqi1iGg88Yvx7scZVAks5qGK6ggaJpZM4IlqS2>
.

goltermann commented:
@smarterclayton can you rebase and get merged?  This appears to be the fix for a P0 (#26036)

smarterclayton committed:
Remove unused deep_copies

smarterclayton commented:
Rebased, with one change.  If `DeepCopy_<localName>` is defined already (in a non-autogenerated file in the package) then the generator will use that method instead of calling into the reflective path.

smarterclayton committed:
Use inline assignment on string alias deep copies

wojtek-t commented:
LGTM - I'm applying a label. We can probably revert it if we decide to push Tim`s approach in 1.3 (which I think we wouldn't).

smarterclayton commented:
@k8s-bot test this issue #26431

smarterclayton commented:
Weird, seems like same problem:

@k8s-bot test this issue #26431

wojtek-t commented:
@smarterclayton - there was some serious issue with this test (I faced this 8 times today). The fix for it has just been merged ~30m ago. Should work now.

smarterclayton commented:
Thanks

yujuhong commented:
@k8s-bot node e2e test this issue: #26431

smarterclayton commented:
@k8s-bot test this issue #IGNORE

thockin commented:
I hate you.  LGTM.

On Wed, Jun 8, 2016 at 8:07 AM, k8s-merge-robot <notifications@github.com>
wrote:

> Merged #26179 <https://github.com/kubernetes/kubernetes/pull/26179>.
>
> —
> You are receiving this because you were mentioned.
> Reply to this email directly, view it on GitHub
> <https://github.com/kubernetes/kubernetes/pull/26179#event-685919190>, or mute
> the thread
> <https://github.com/notifications/unsubscribe/AFVgVBKN1KvI4AFiI3F9_o-gLhDszinhks5qJtqngaJpZM4IlqS2>
> .
>
